lua/neo-tree/sources/diagnostics/lib/items.lua

```diff
@@ -132,6 +132,11 @@ M.get_diagnostics = function(state)
             end
           end
           table.sort(diag_items, sort_func)
+          for i = #diag_items, 1, -1 do
+            if diag_items[i + 1] and diag_items[i].id == diag_items[i + 1].id then
+              table.remove(diag_items, i)
+            end
+          end
```

Comment:
This solution assumes that nodes with the same id lie together after the items have been sorted, but there are a few situations where this is not necessarily the case. The most common case in practice would be if there was another item which is considered equal to the duplicated item with regard to `sort_func`. Since `table.sort` is not a stable sort, this item might appear between the duplicated items after sorting.

I think a better approach would be recording the id when a diagnostic item is created and only adding it to the list of diagnostics if the id has not been encountered. This would be best achieved with a table of the form `{ <id> = true, ... }`.

Reply:
Oh, hadn't considered that it could sort "too much"

You mean `{<id> = <item>, ... }`, right? so using the "hash table side" of Lua tables, to de-duplicate on insertion? it does feel like _the Lua Way_

Reply:
Actually that could also work if you want to choose the duplicate with the highest severity or something similar (since severity does not affect the id), but no, I did mean `{ <id> = true, ... }`. I've heard it referred to as a Lua set. You would then add to the set with `encountered[<id>] = true` and could check if an id has been encountered using `if encountered[<id>] then ...`. Very _the Lua way_.